### ISSUE TITLE
Support boot in EFI mixed mode

### ISSUE THREAD

**Issue by mirh** [systemd/systemd#17056]
**Is your feature request related to a problem? Please describe.**
Booting 64-bit kernels under a 32-bit UEFI firmware (and viceversa perhaps?).

**Describe the solution you'd like**
This probably [used](https://lkml.org/lkml/2014/3/4/242) to be [outside](https://www.slideshare.net/mattjfleming/bradlug-efimixed) the scope of this project, but as of linux 5.8 AFAIU it should be in the neighborhood of possible. 
https://patchwork.kernel.org/cover/11380511/
https://patchwork.kernel.org/cover/11368503/

**Describe alternatives you've considered**
Well, I tried to load a x64 kernel from systemd-bootia32.efi, but everything I got was `Failed to execute Arch Linux (\vmlinuz): Unsupported`.

**Comment by poettering** [comment]
sd-boot doesn't actually boot any kernels, it just launches EFI binaries. Hence, if what you are looking for can be implemented as EFI binary, then things should just work. i.e. if you can build a 64bit kernel that looks like a 32bit UEFI binary, then it should just work.

**Comment by mirh** [comment]
It is my understanding that we still aren't at a level that you could just boot a 64-bit binary directly from 32-bit firmware. On the other hand I think minimal changes would be now required (as opposed to the handover protocol.. which anyway I just discovered is already kind of [supported](https://github.com/systemd/systemd/pull/11772)?)

Maybe @ardbiesheuvel could comment on what the long game is.

**Comment by ardbiesheuvel** [comment]
Yes, recent 64-bit kernels can easily be launched by 32-bit firmware on x86, provided that the LoadImage() boot service permits loading of X64 images (which is usually the case).

The call to StartImage() will fail in this case, but the PE/COFF image will have a .compat section, which contains the machine type and the entrypoint address (which in this particular case, points to a IA32 compatible entrypoint)

Please refer to https://github.com/tianocore/edk2/commit/d55cfdc51f93b6622c3780d4cc2ba2769c154a15 for an example implementation. In this case, it hooks into the image loading logic, so X64 images can be started from the IA32 UEFI shell as well.

**Comment by poettering** [comment]
OK, so I guess judging by @ardbiesheuvel' comments we can close this as the support is in the kernel already and sd-boot doesn't implement the any boot protocols natively anyway, and exclusively relies on the kernel's own EFI/PE wrapping?

**Comment by mirh** [comment]
Mhh no? What I understood from the comment is that indeed you to make slight adgu
(also, according to the PR I linked, I thought that boot was not having a special regard for linux protocols?)

**Comment by poettering** [comment]
> Mhh no? What I understood from the comment is that indeed you have to make some slight adjustments to the loading mechanism.
> (also, according to the PR I linked, I thought that boot was now having special regards for linux protocols?)

That PR is about the EFI stub, not systemd-boot.

systemd provides two EFI components: systemd-boot (which is a boot menu only, all it can do is enumerate boot menu items found on disk, and show them in the menu, and eventually load them as regular EFI binaries, that's all. it has no loader capabilities on its own), and an EFI stub which you can link into the linux kernel if you like, which can display a boot splash pre-initialization and passes some metadata down to userspace that is useful. The stub is part of the kernel binary ultimately, and preps stuff for transition into linux mode, that's what the referenced PR is about.

Anyway, if there's stuff to make work what you want, then by all means send a patch, but otherwise I doubt anyone will work on this.

**Comment by ardbiesheuvel** [comment]
On Fri, 16 Oct 2020 at 18:16, Lennart Poettering
<notifications@github.com> wrote:
>
> Mhh no? What I understood from the comment is that indeed you have to make some slight adjustments to the loading mechanism.
> (also, according to the PR I linked, I thought that boot was now having special regards for linux protocols?)
>
> That PR is about the EFI stub, not systemd-boot.
>
> systemd provides two EFI components: systemd-boot (which is a boot menu only, all it can do is enumerate boot menu items found on disk, and show them in the menu, and eventually load them as regular EFI binaries, that's all. it has no loader capabilities on its own), and an EFI stub which you can link into the linux kernel if you like, which can display a boot splash pre-initialization and passes some metadata down to userspace that is useful. The stub is part of the kernel binary ultimately, and preps stuff for transition into linux mode, that's what the referenced PR is about.
>
> Anyway, if there's stuff to make work what you want, then by all means send a patch, but otherwise I doubt anyone will work on this.
>
>

Mixed mode support requires some changes on the boot loader side,
simply because the binary in question (the x86_64 kernel) is not a
firmware native binary (i386 in the case of x86 mixed mode).

Out-of-tree GRUB uses the EFI handover protocol for this, but that is
simply a hybrid between a x86 specific bootloader (using struct
bootparams from the Linux tree, and calling a non-EFI entrypoint at a
known fixed offset into the binary), which relies on separate image
loading functionality too.

So what we provided instead is a generic mixed-mode entry point, which
is fairly trivial to implement in a generic EFI loader app: it uses
the LoadImage() boot service (which allows X64 images to be loaded on
IA32 firmware implementations) to get the image into memory, but
because StartImage() cannot be called, the loader app needs to parse
the PE/COFF headers to find the .compat section, which contains a set
of {arch type, entrypoint} tuples. On mixed mode capable X64 images,
this will contain an entry for IA32 which can be invoked in the
ordinary way, i.e., passing a handle and a EFI system table pointer.

This is still on the wishlist for upstream GRUB, but nobody needs it
so desperately that they are willing to work on it. As I mentioned
before, we have a working implementation for EDK2, including one for
the UEFI shell.

**Comment by lorn10** [comment]
Hi all, - just out of curiosity, was there any progress in this matter? Is sytemd able to handle this EFI mixed mode topic (as of 2022) properly? 

Note, I'm now thinking this particular in conjunction with the [rEFInd](https://www.rodsbooks.com/refind/) boot manager and not GRUB. 

Whatever, I guess that also rEFInd most likely needs some additional modifications to allow a mixed Mode EFI operation in conjunction with sytemd or the bare Linux kernel (EFISTUB).

**Comment by ardbiesheuvel** [comment]
> Hi all, - just out of curiosity, was there any progress in this matter? Is sytemd able to handle this EFI mixed mode topic (as of 2022) properly?
> 
> Note, I'm now thinking this particular in conjunction with the [rEFInd](https://www.rodsbooks.com/refind/) boot manager and not GRUB.
> 
> Whatever, I guess that also rEFInd most likely needs some additional modifications to allow a mixed Mode EFI operation in conjunction with sytemd or the bare Linux kernel (EFISTUB).

In a nutshell, whichever link in the boot chain calls LoadImage() on the kernel image needs to deal with StartImage() potentially returning a failure, in which case the compat entrypoint needs to be invoked instead (provided that one exists for the arch in question)

**Comment by medhefgo** [comment]
I already have some WIP code for this that should be ready soonish.

But today I found out that it already seems to work with systemd-stub (on x86). **Can anyone with *real* hardware try making a unified kernel image with the ia32 systemd-stub (from debian for example) and a 64bit kernel with mixed-mode support to confirm this works?**

@ardbiesheuvel The systemd-stub currently uses the (deprecated) efi handover protocol on x86 to support booting older kernels. Is there a 32bit efi handover entry point? Aka, is ia32 systemd-stub booting a 64bit kernel that way supported or just some fluke from running this under qemu?
The boot protocol doc says `EFI handover protocol entry point which is hdr->handover_offset bytes from the beginning of `startup_{32,64}`, but no mention of where that is. The stub uses `code32_start` (+ 512 on x64), see [linux_x86.c](https://github.com/systemd/systemd/blob/2f7a0648cdf73d30bae6d0ab627abe77c4ba90ec/src/boot/efi/linux_x86.c#L109).

Btw, I do make sure to uninstall the coff emulation from ovmf in systemd-boot to make sure it doesn't do compat entry lookup for me.

Ultimately, I find the way the kernel does mixed mode to be rather handwavy. For fun I tried running a 32bit kernel on my 64bit efi and `LoadImage` succeeds while `StartImage` fails. But an aarch64 kernel will fail with `EFI_UNSUPPORTED` in `LoadImage`. This makes no sense to me and hoping that all ia32 firmware behaves that way sounds like a really big assumption to me…

**Comment by ardbiesheuvel** [comment]
> @ardbiesheuvel The systemd-stub currently uses the (deprecated) efi handover protocol on x86 to support booting older kernels. Is there a 32bit efi handover entry point? Aka, is ia32 systemd-stub booting a 64bit kernel that way supported or just some fluke from running this under qemu? The boot protocol doc says `EFI handover protocol entry point which is hdr->handover_offset bytes from the beginning of `startup_{32,64}`, but no mention of where that is. The stub uses `code32_start` (+ 512 on x64), see [linux_x86.c](https://github.com/systemd/systemd/blob/2f7a0648cdf73d30bae6d0ab627abe77c4ba90ec/src/boot/efi/linux_x86.c#L109).
> 

If the kernel is built with CONFIG_EFI_MIXED, there will be a 32-bit EFI handover protocol entrypoint at offset 0x190.

> Btw, I do make sure to uninstall the coff emulation from ovmf in systemd-boot to make sure it doesn't do compat entry lookup for me.
> 
> Ultimately, I find the way the kernel does mixed mode to be rather handwavy. For fun I tried running a 32bit kernel on my 64bit efi and `LoadImage` succeeds while `StartImage` fails. But an aarch64 kernel will fail with `EFI_UNSUPPORTED` in `LoadImage`. This makes no sense to me and hoping that all ia32 firmware behaves that way sounds like a really big assumption to me…

AArch64 is radically different. ARM cannot run 32-bit code and 64-bit code at the same exception level, so the concept of EFI mixed mode does not exist on ARM.

**Comment by ardbiesheuvel** [comment]
> > AArch64 is radically different. ARM cannot run 32-bit code and 64-bit code at the same exception level, so the concept of EFI mixed mode does not exist on ARM.
> 
> I was referring to loading it on x86. It makes no sense to me that a real x86_64 firmware would happily load 32bit code but not launch it, but at the same completely refuse to _load_ aarch64 code.

What do you mean by 'real x86_64 firmware'? The vast majority of production EFI firmware deployed on x86 hardware is based on EDK2 source code, even though some BIOS vendors will lie and say it was written from scratch based on the UEFI and PI specifications.

I don't see how the AArch64 situation has any bearing on that.

**Comment by ardbiesheuvel** [comment]
> > > AArch64 is radically different. ARM cannot run 32-bit code and 64-bit code at the same exception level, so the concept of EFI mixed mode does not exist on ARM.
> > 
> > 
> > I was referring to loading it on x86. It makes no sense to me that a real x86_64 firmware would happily load 32bit code but not launch it, but at the same completely refuse to _load_ aarch64 code.
> 
> What do you mean by 'real x86_64 firmware'? The vast majority of production EFI firmware deployed on x86 hardware is based on EDK2 source code, even though some BIOS vendors will lie and say it was written from scratch based on the UEFI and PI specifications.
> 
> I don't see how the AArch64 situation has any bearing on that.

Btw, the point of mixed mode is loading 64-bit code on 32-bit firmware, not the other way around.

The cross type compatibility of X64 code on IA32 firmware was introduced into EDK2 in 2007: https://github.com/tianocore/edk2/commit/a7ed1e2ed5305d2b1e3ed2ccfe7f9d0cab564106.

**Comment by medhefgo** [comment]
> Btw, the point of mixed mode is loading 64-bit code on 32-bit firmware, not the other way around.

I was just doing that to poke around. But I would not be surprised someone out there would wanna do that for real.

Ultimately, I was just voicing my thoughts on the way this is implemented in firmware and relying on it to be stretching it. But if that is what the kernel does, we'll support it.

**Comment by ardbiesheuvel** [comment]
> > Btw, the point of mixed mode is loading 64-bit code on 32-bit firmware, not the other way around.
> 
> I was just doing that to poke around. But I would not be surprised someone out there would wanna do that for real.
> 

32-bit code cannot call 64-bit firmware services, even on x86, so even if the image could be started, there is no meaningful way it could interact with the firmware after that, which kind of defeats the point of EFI boot.
 
> Ultimately, I was just voicing my thoughts on the way this is implemented in firmware and relying on it to be stretching it. But if that is what the kernel does, we'll support it.

The alternative is to implement the EFI handover protocol, as it is not going away. But that means introducing a lot of x86/Linux specific logic (struct bootparams, where to put the command line or initrd in memory, etc).With the compat entrypoint and LoadFIle2 based initrd loading, everything is generic and supported across all EFI architectures that Linux supports.

**Comment by medhefgo** [comment]
Please test the attached PR on real hardware. Add `-Defi-cflags=-m32` to the meson call; not that this requires a ia32 gnu-efi to be available (works on stock debian/ubuntu).

**Comment by ardbiesheuvel** [comment]
I don't have access to any real hardware at the moment, but this code works fine for me booting a mixed mode kernel from 32-bit OVMF running under x86_64 QEMU/kvm.

Tested-by: Ard Biesheuvel <ardb@kernel.org>

Thanks a lot for working on this.

@jwrdegoede in case you have some spare bandwitdh and want to give this a spin?

**Comment by medhefgo** [comment]
> Not sure I understand the difference. I just created a .conf file to load a bzImage kernel image that has the .compat entrypoint.

That means you tested the regular boot entries, which is good, cause it actually exercises the new code. That leaves the question whether your OVMF has the coff emulation support or not.

**Comment by ardbiesheuvel** [comment]
> > Not sure I understand the difference. I just created a .conf file to load a bzImage kernel image that has the .compat entrypoint.
> 
> That means you tested the regular boot entries, which is good, cause it actually exercises the new code. That leaves the question whether your OVMF has the coff emulation support or not.

I tested both with and without.

**Comment by medhefgo** [comment]
Btw, I noticed that a doctored image with `type > 1` and `size == 0` could cause an infinite loop in OVMF: https://github.com/tianocore/edk2/blob/8a576733162bb72afb4d1eb3012b0aef8d265018/OvmfPkg/CompatImageLoaderDxe/CompatImageLoaderDxe.c#L84

**Comment by lorn10** [comment]
@medhefgo I have available an Apple **iMac5,1** computer which has a **32bit EFI** firmware. It is running Kubuntu 20.04 LTS (currently through CSM) which is a 64bit distro.

So if needed I could test a bootable CD that does the 32bit to 64bit transition mentioned here. However, I am not able to do all the code stuff. :wink:

**Comment by medhefgo** [comment]
@lorn10 Easiest way to test is by creating a fat32 format usb drive, extract the [arch iso](https://archlinux.org/download/) into it and then put [BOOTia32.EFI](https://c.web.de/@335398652035989738/Xnwh3f_gQ7ymLFNMfbyEpA) (or compile it yourself) into `EFI/BOOT/` on the usb drive. Then just boot from it and make sure it says `ia32` when pressing `p` in the boot menu. Also, if booting works, make sure it says `32` in `/sys/firmware/efi/fw_platform_size`.

**Comment by lorn10** [comment]
Thanks for this information. Okay, I think the underlying principle works so far. I get the Arch Linux boot screen with the three options. But when the first one is selected it goes some time and then it is written (in red): `Error loading \arch\boot\x86_64\vmlinuz-linux: Unsupported`

This is true for a "direct-load" from the stock Apple EFI loader as well from rEFInd. 

PS I think we are now off-topic and should discuss this "somewhere other". (This seems to be an Arch problem.) :wink:

**Comment by mirh** [comment]
Arch is built with CONFIG_EFI_MIXED=y. 
So the thing not booting (even though the error seems already different than the one I was getting on my T100TA) is quite in-topic.

**Comment by medhefgo** [comment]
@lorn10 Thanks for testing, this is quite valuable info.

@ardbiesheuvel This is what I was fearing: it seems the assumption that firmware supports loading ia32 images in `LoadImage` doesn't always hold. :P

I can probably work around this by leveraging falling back on sd-stub code, which should work regardless. Will look into that tomorrow.

**Comment by ardbiesheuvel** [comment]
> @lorn10 Thanks for testing, this is quite valuable info.
> 
> @ardbiesheuvel This is what I was fearing: it seems the assumption that firmware supports loading ia32 images in `LoadImage` doesn't always hold. :P
> 
> I can probably work around this by leveraging falling back on sd-stub code, which should work regardless. Will look into that tomorrow.

At which stage is systemd-bootia32.efi being used in this particular example?

**Comment by ardbiesheuvel** [comment]
> > @lorn10 Thanks for testing, this is quite valuable info.
> > @ardbiesheuvel This is what I was fearing: it seems the assumption that firmware supports loading ia32 images in `LoadImage` doesn't always hold. :P
> > I can probably work around this by leveraging falling back on sd-stub code, which should work regardless. Will look into that tomorrow.
> 
> At which stage is systemd-bootia32.efi being used in this particular example?

Never mind. It's your special build of bootia32.efi, right?

So yes, in that case, you are right that LoadImage() may reject X64 images on IA32 firmware. These Apple machines are running EFI 1.10 forks, so they are probably rather ancient, but OTOH, they are one of the main uses for mixed mode.

So the only solution to cover these systems as well is reimplementing image loading from scratch, similar to how Shim/Grub do this.

**Comment by medhefgo** [comment]
@mirh Btw, can you give the above build a try? Based on the error message from your initial post, it should work.

**Comment by lorn10** [comment]
And maybe @1nikolas is also interested in trying this out (of course just in case he still has his old Lenovo "BayTrail" laptop). :+1:

**Comment by 1nikolas** [comment]
> And maybe @1nikolas is also interested in trying this out (of course just in case he still has his old Lenovo "BayTrail" laptop). 👍

thanks for the tag. I used the bootia32 from @medhefgo with a year old arch iso (that's what I had laying around) and it indeed works (sorry for the crap photos)

![image](https://user-images.githubusercontent.com/30593419/154864169-15368a1a-eae2-4460-949d-dca8fa70e043.png)
![image](https://user-images.githubusercontent.com/30593419/154864176-8cf803bb-b642-4d86-ba18-94fd646b59c3.png)


For reference my laptop is [this](https://www.amazon.co.uk/Lenovo-Ideapad-100s-11IBY-1-33GHz-Windows/dp/B01NCBCSA0)

**Comment by lorn10** [comment]
@medhefgo Just for the sake of completeness. I don't know if the following has some relevance here but there exist regarding the Apple EFI 1.10 forks another quite annoying behavior in regard to the vbios. That one is on those early mac models usually only available for a short time on startup (power-on). Other boot-loaders like rEFInd "extend" that period and keep this information for the Linux kernel (in EFISTUB boot config) which is then able to get the vbios information via "[ACPI VFCT vbios fetch](https://bugs.freedesktop.org/show_bug.cgi?id=26891#c32)". When this information is missing, Apple models with Radeon GPUs will fail booting with message `*Error* No UMS support in Radeon module!`The only alternative way to boot up such systems is with `nomodeset` parameter to get at least some basal "save graphics" functionality.

Long story short, it would be really useful if this implementation here could do (in that vbios matter) the same as rEFInd and preserve that critical information for the Linux kernel. However, I have no clue if this can work at all in an EFI mixed mode scenario.

**Comment by ardbiesheuvel** [comment]
> Long story short, it would be really useful if this implementation here could do (in that vbios matter) the same as rEFInd and preserve that critical information for the Linux kernel. However, I have no clue if this can work at all in an EFI mixed mode scenario.

If booting via the EFI stub is sufficient to get the correct behavior regarding the VFCT table, using the compat entrypoint should work, provided that LoadImage() can be emulated, and provides an appropriate EFI_LOADED_IMAGE instance.

It is not entirely clear to me how a ACPI table such as VFCT is affected by this, though. Is it omitted when booting via the CSM? Why does the EFI stub make a difference here?

**Comment by medhefgo** [comment]
@lorn10 I've updated the efi binary (same link), please give it a try.

Regarding the vbios: if all is needed is the stub loader, than there shouldn't be anything necessary on our side.

@ardbiesheuvel  Does the kernel save the vbios with efi handover protocol too?

**Comment by ardbiesheuvel** [comment]
> @lorn10 I've updated the efi binary (same link), please give it a try.
> 
> Regarding the vbios: if all is needed is the stub loader, than there shouldn't be anything necessary on our side.
> 
> @ardbiesheuvel Does the kernel save the vbios with efi handover protocol too?

I don't see how any of this is related to EFI - perhaps the Apple CSM interacts weirdly with the Radeon EFI driver. The kernel just looks af the VFCT ACPI table if it exists, and this is completely unrelated to the EFI stub afaict.

**Comment by lorn10** [comment]
@medhefgo Unfortunately the iMac is now simply freezing when I enter the main "Arch Linux install" option. No error message is displayed, just a black screen. I have also tried the `nomodeset` `enforcing=0` `xforcevesa` and `vga=791` parameter but with same result. (These options are necessary for example in conjunction with [GParted](https://gitlab.gnome.org/GNOME/gparted/-/issues/167) to boot on "vbios-restricted" Apple computers via the native **EFI mode** with minimal "save graphics").

@ardbiesheuvel As you most likely know, Apple computers have some form of a "hybird vbios" which includes a legacy and a EFI vbios image, see example below. I am talking here about the EFI vbios part not the legacy one. At "vbios-restricted" Apple models the EFI vbios information is only available at the power-on. (This stands in contrast to booting via CSM, where the legacy vbios is always exposed.) The "VFCT patch" from David L and Alex Deucher makes it possible for the kernel to get the necessary information. Maybe we should ping on of them, - at least Stefan Dösinger should be reachable on GitHub. I am here just "the guy with an iMac5,1 computer". :wink: 

Whatever, so far I know this is only possible when booting the bare Linux kernel (EFISTUB) from the stock Apple EFI or rEFInd. This does not work with GRUB.

For reference a vbios dump information example, **Radeon HD 4850** (MXM card) out of an **iMac11,1** computer:
```
kubuntu@kubuntu: .../Downloads/atiflash_linux_4.25$ sudo ./atiflash -ai
Adapter  0    (BN=01, DN=00, PCIID=944A1002, SSID=00B5106B)
    Asic Family        :  RV770/M98      
    Flash Type         :  PM25LV010   (128 KB)
    Product Name       :  113-B91101-009 M98 GDDR3 256bit 503e/850m 
    Bios Config File   :  13B91101.009   
    Bios P/N           :  113-B91101-009 
    Bios Version       :  011.022.007.004.034277
    Bios Date          :  08/28/09,19:19:2
    ROM Image Type     :  Hybrid Images
    ROM Image Details  :  
        Image[0]: Size(63488 Bytes), Type(Legacy Image)
        Image[1]: Size(46080 Bytes), Type(EFI Image)
```

**Comment by medhefgo** [comment]
Try booting booting with "loglevel=6 earlyprintk=efi efi=debug" (arch kernel is silent by default).

**Comment by lorn10** [comment]
Indeed, a new message appears (after quite some time):
`EFI stub: Loaded initrd from LINUX_EFI_INITRD_MEDIA_GUID device path`

And then it freezes, - I had to hard power-off the iMac via the power-button. (Same as before.)

**Comment by lorn10** [comment]
A new addition: I was really for the first time able to install a **32bit** distro in **native EFI mode** with full graphics support (acceleration), at an **iMac5,1**. :+1: 

As mentioned, this newer worked for me before. (This worked in EFI mode only with the above listed "save graphics" parameter.)

I simply installed the latest MX-21 "Wildflower" release (with kernel 5.10) onto an external SSD but **without any GRUB support**. (This was kindly offered by the installer.) After that I have booted into Kubuntu 20.04 LTS and copied the MX Linux `vmlinuz-5.10.0-9-686-pae` to `vmlinuz-5.10.0-9-686-pae.efi`. Then I rebooted again and indeed there was a new "Debian" option avaiable in rEFInd. I chose that, and for the first moment it looked that the whole system was frozen again. But to my big surprise it suddenly continued and I reached the login screen.

The following information may be here of special interest. More can be found in the full dmesg log in the attachment.
```
[    4.090571] [drm] radeon kernel modesetting enabled.
[    4.090648] radeon 0000:01:00.0: vgaarb: deactivate vga console
[    4.091051] [drm] initializing kernel modesetting (RV530 0x1002:0x71C5 0x106B:0x0080 0x00).
[    4.091118] radeon 0000:01:00.0: Invalid PCI ROM header signature: expecting 0xaa55, got 0xa3a3
[    4.091144] radeon 0000:01:00.0: Invalid PCI ROM header signature: expecting 0xaa55, got 0x2323
[    4.092810] ATOM BIOS: M56P
[    4.092830] [drm] GPU not posted. posting now...
[    4.103406] [drm] GPU not posted. posting now... 
```
```
[    4.125681] radeon 0000:01:00.0: firmware: direct-loading firmware radeon/R520_cp.bin
```
[dmesgiMac5,1efi(32bit).txt](https://github.com/systemd/systemd/files/8125819/dmesgiMac5.1efi.32bit.txt)
[glxinfoiMac5,1efi(32bit).txt](https://github.com/systemd/systemd/files/8125820/glxinfoiMac5.1efi.32bit.txt)

**Comment by mirh** [comment]
Newer kernels may include [this fix](https://gitlab.freedesktop.org/xorg/driver/xf86-video-nouveau/-/issues/211#note_477672). 
Anyhow, if you got your system to at least start reliably with other means, now you could get back in topic of testing systemd-boot.

**Comment by medhefgo** [comment]
> Indeed, a new message appears (after quite some time): `EFI stub: Loaded initrd from LINUX_EFI_INITRD_MEDIA_GUID device path`
> 
> And then it freezes, - I had to hard power-off the iMac via the power-button. (Same as before.)

This means we're already inside the kernel stub entry point and it's freezing up for some reason. It likely doesn't like how it was set up/loaded (wouldn't know quite what, though. Possibly because no PE relocations were done).

Updated systemd-boot with some debug logs: [BOOTia32.efi](https://c.web.de/@335398652035989738/Xnwh3f_gQ7ymLFNMfbyEpA)
Here's a [unified kernel image](https://c.web.de/@335398652035989738/72Pic2xoRJ6zL69S7BYGPg) that is more likely to succeed. Put it in `\EFI\Linux\` (on the arch USB drive) and try booting the new entry.

**Comment by lorn10** [comment]
Thanks :+1:  and sorry for my late feedback. (Please ping me additionally the next time).

Okay, there was no `\EFI\Linux\` location present so I made a new `Linux` folder there and put in the `arch.efi`.

The new `BOOTia32.EFI` gives me when I select _Arch Linux install medium (x86_64,  UEFI)_ after quite some time a red `Calling kernel_entry` message.

When I select the first _Arch Linux_ entry then the iMac just reboots immediately.

Finally when I select directly the `arch.efi` from rEFInd then I get the following output:
```
systemd-stub@0xADC0E000,0xADC1B000
Calling linux_efi_handover
```
(The second line is in orange.)

**Comment by medhefgo** [comment]
> The new `BOOTia32.EFI` gives me when I select _Arch Linux install medium (x86_64, UEFI)_ after quite some time a red `Calling kernel_entry` message.

This suggests that `LoadImage` does work, while it previously didn't? I am confused.

Here's an updated version with a different approach (you should delete the old arch.efi): [BOOTia32.efi](https://c.web.de/@335398652035989738/Y2Sl0VC3SHWR-XLNWum0sQ) [arch-x86.efi](https://c.web.de/@335398652035989738/1qRzPgU-R3iebdr0AJWhBQ) [arch-generic.efi](https://c.web.de/@335398652035989738/cHfss2FhS1OurYWdV0tYsw)

@lorn10 Please give booting the regular arch install boot entry and both the unified loader versions a try (if needed, add any extra cmdline args).

**Comment by lorn10** [comment]
A further short side note, - maybe the following has here some relevance.

It looks that the **32bit** Apple EFI 1.10 forks have some issues with GNU-EFI which are not present with TianoCore. I quote a [comment](https://www.rodsbooks.com/refind/installing.html) from **Roderick W. Smith**:

> Warning: I know of one bug with the version of rEFInd built with GNU-EFI: On my 32-bit Mac Mini, the filesystem drivers hang when launched. This can render the system unbootable until you bypass rEFInd. This bug does not manifest when running the same binaries under a 32-bit VirtualBox, and I've never run into it on any 64-bit system (including a 64-bit MacBook Air). Debugging suggests that a function is being entered mid-function, which implies a bug in the EFI or in the development tools. In any event, the bottom line is to not use the [GNU-EFI compiled builds from the rEFInd] PPA on a 32-bit Mac.

So it looks that it makes quite a difference _how_ the efi files are build. :wink:  @medhefgo

**Comment by medhefgo** [comment]
The refind fs driver issue shouldn't be an issue for sd-boot as we simply do not use them.

As the log message `EFI stub: Loaded initrd from LINUX_EFI_INITRD_MEDIA_GUID device path` shows, we are already running inside of the kernel (that is the kernel stub talking, not ours). I have no idea why it would freeze up and I've run out of ideas as to how to start the kernel differently.
The last binary I provided simply patches in the ia32 machine type and ia32 stub entry point into the pe binary and then uses the fimrware provided `LoadImage`. And it seems to work on your machine (and definitely does in QEMU), after all, the kernel stub is running and even fetching the initrd we provide, but then freezes up on something else. Maybe it really is a kernel bug?

Ultimately, I am inclined to just give up on supporting non-UEFI firmware…

**Comment by lorn10** [comment]
Yes, even if it's unfortunate, I can understand when these old and bizarre 32bit EFI 1.10 Apple systems are not supported by this project here. Maybe it's simply not possible. Maybe these systems are in the end effectively incapable to work according the EFI "mixed mode principle". Whatever, or it could also be that someday someone will find out a workaround for this. :wink: 

@medhefgo Thanks for yours efforts in this matter! :+1: And just for the record, are these efi files here made with [TianoCore](https://www.tianocore.org/)? I think that information might be eventually useful sometime (in the future) for others.

**Comment by medhefgo** [comment]
@lorn10 The 5.17 kernel has a fix that might may make this work now. You can extract `/usr/lib/modules/5.17.0-arch1-1/vmlinuz` from https://archlinux.org/packages/testing/x86_64/linux/download and then replace the one on the USB stick at `arch/boot/x86_64/vmlinuz` (or something like that) and then try booting the regular entry again.

**Comment by lorn10** [comment]
Thanks for this additional information. Well, the 32bit Apple EFI 1.10 variant seems to be unfortunately rally somewhat "jinxed". It freezes also with that 5.17 kernel based vmlinuz file:

```
Successfully patched kernel image.
EFI stub: Loaded initrd from LINUX_EFI_INITRD_MEDIA_GUID device path
```
Whatever, maybe the Mac OS tweaking guys from the [Macosxbootloader](https://piker-alpha.github.io/macosxbootloader/) project @Piker-Alpha and @mhaeuser have some final ideas what we may not have been considered here. It looks that they are really able to boot a 64bit Mac OS kernel from a 32bit Apple EFI. :wink:

**Comment by mhaeuser** [comment]
@lorn10 I haven't been involved much with that project, but please remember two things:

1) EFI 1.10 era firmwares were Wild West and Apple remained to maintain an awkward "somewhat compatible" firmware design to please Windows, but had and has various deviations from the specs. I do not know anyone who really deals with Macs this old and I do not believe there is much "safe" knowledge about their firmware at that point in time. This probably needs careful debugging (I didn't see a comment of anyone actually investigating the crash with hardware-assisted debugging and I'm not sure how easy it is with old Macs).

2) This is not a unique solution of the project. macosxbootloader is mostly reversed from Apple's own efiboot and macOS has been designed with mixed-arch in mind. What I would do is reverse the shims in macOS to call 32-bit EFI functions and check for unexpected operations (e.g. may they have used some deviated calling conventions with stricter requirements?) and record all EFI calls from the Linux side (are they even provided by Apple EFI?), services and protocols. I would not be surprised if Apple unintentionally advertised some function in a services table or protocol to be available when it isn't.

**Comment by mirh** [comment]
Cool, thanks!
Btw here's some quirks I could find for old macs booting
https://github.com/rhboot/grub2/commit/fc8584825cb513ffaca12585657b1682ef79a3e5
https://github.com/rhboot/grub2/commit/7fae005102ffab0148219a26df8aa80d2110627e
https://github.com/rhboot/grub2/commit/f9811a92e6d4c0579f989f3d6dcc66e8fb35cb09

Maybe @mjg59 knows some extra idiot balls, even though I think he just focused on the kernel aspect of the thing.

**Comment by mjg59** [comment]
I don't think any of those are relevant to this case. The best thing you can probably do here is add additional printks to the stub and try to narrow down where it's hanging (you'll need earlyprintk enabled to see it) - if we're executing code in the stub then this is likely to be a kernel bug rather than anything else, but mixed mode does complicate things.

**Comment by ardbiesheuvel** [comment]
Implementing mixed mode using the .compat entrypoint involves either relying on LoadImage() to load the non-native image as usual, or implementing a PE/COFF loader from scratch. Given that EDK2's LoadImage() behaves as desired in this respect, the EDK2 version of x86  mixed mode boot uses the former, simpler approach, and as I understand it, this is the approach taken by systemd-boot as well.

We could investigate a bit further how exactly IA32 EFI implementations on Mac machines behave when presented with X64 images, and work around any shortcomings, but I'm not convinced this is worth anyone's time tbh.

**Comment by tlhunter** [comment]
I'm not sure if this is of use but I seem to be having the same issue outlined in this thread. Specifically I'm using the following hardware:

```
Apple iMac "Core 2 Duo" 2.66 20" (Early 2008)
Identifiers: Early 2008 - MB324LL/A - iMac8,1 - A1224 - 2210
```

This happens both with the whatever Apple uses by default to boot and also with rEFInd. The symptom that I get is that upon booting, after the GRUB menu is displayed, and assuming I've disabled `quiet` and `splash`, the following message displayed for approximately 4 minutes:

```
efi stub: loaded initrd from LINUX_EFI_INITRD_MEDIA_GUID device path
```

Once that time passes I then see the Kernel logs and then the boot screen. This happens with different versions of Kernel 6.8 and a couple different distros.
